Docs: Improve docs for plugin authentication (#33325)
grafana/grafana: docs/sources/developers/plugins/add-authentication-for-data-source-plugins.md

```ts
interface Props extends DataSourcePluginOptionsEditorProps<MyDataSourceOptions, MySecureJsonData> {}
```

You can access the value of the secret from the `options` prop inside your `ConfigEditor` until the user saves the configuration. When the user saves the configuration, Grafana clears the value. After that, you can use the `secureJsonFields` to determine whether the property has been configured.

```ts
const { secureJsonData, secureJsonFields } = options;
const { apiKey } = secureJsonData;
```

To securely update the secret in your plugin's configuration editor, update the `secureJsonData` object using the `onOptionsChange` prop.

```ts
const onAPIKeyChange = (event: ChangeEvent<HTMLInputElement>) => {
  onOptionsChange({
    ...options,
    secureJsonData: {
      apiKey: event.target.value,
    },
  });
};
```

Next, define a component that can accept user input.

```ts
<Input
  type="password"
  placeholder={secureJsonFields?.apiKey ? 'configured' : ''}
  value={secureJsonData.apiKey ?? ''}
  onChange={onAPIKeyChange}
/>
```

> Finally, if you want the user to be able to reset the API key, then you need to set the set it to `false` in the `secoreJsonFields` object.

Author: ```suggestion
Finally, if you want the user to be able to reset the API key, then you need to set property to `false` in the `secureJsonFields` object.
```

Now that users can configure secrets, the next step is to see how we can add them to our requests.

## Authenticate using the data source proxy

Once the user has saved the configuration for a data source, any secret data source configuration will no longer be available in the browser. Encrypted secrets can only be accessed on the server. So how do you add them to you request?

The Grafana server comes with a proxy that lets you define templates for your requests. We call them _proxy routes_. Grafana sends the proxy route to the server, decrypts the secrets along with other configuration, and adds them to the request before sending it off.

> **Note:** Be sure not to confuse the data source proxy with the [auth proxy](TODO). The data source proxy is used to authenticate a data source, while the auth proxy is used to log into Grafana itself.

### Add a proxy route to your plugin

To forward requests through the Grafana proxy, you need to configure one or more proxy routes. A proxy route is a template for any outgoing request that is handled by the proxy. You can configure proxy routes in the [plugin.json](https://grafana.com/docs/grafana/latest/developers/plugins/metadata/) file.

> 1. Add the route to plugin.json. Note that you need to reload the Grafana server every time you make a change to your plugin.json file.

Author: ```suggestion
1. Add the route to plugin.json. Note that you need to restart the Grafana server every time you make a change to your plugin.json file.
```

To authenticate using OAuth 2.0, add a `tokenAuth` object to the proxy route definition. If necessary, Grafana performs a request to the URL defined in `tokenAuth` to retrieve a token before making the request to the URL in your proxy route. Grafana automatically renews the token when it expires.

Any parameters defined in `tokenAuth.params` are encoded as `application/x-www-form-urlencoded` and sent to the token URL.

```json
{
  "routes": [
    {
      "path": "api",
      "url": "https://api.example.com/v1",
      "tokenAuth": {
        "url": "https://api.example.com/v1/oauth/token",
        "params": {
          "grant_type": "client_credentials",
          "client_id": "{{ .SecureJsonData.clientId }}",
          "client_secret": "{{ .SecureJsonData.clientSecret }}"
        }
      }
    }
  ]
}
```

## Authenticate using a backend plugin

While the data source proxy supports the most common authentication methods for HTTP APIs, using proxy routes has a few limitations:

> - Proxy routes only support HTTP

Reviewer: Proxy also supports HTTPS, right? I know what you mean here, but I wonder if saying only "HTTP" might lead people to incorrectly think that they need to develop a backend plugin if they want to use https.

Author: ```suggestion
- Proxy routes only support HTTP/HTTPS
```

Good point! How about this?

Reviewer: looks great :)

> - Proxy routes don't support custom token authentication

Author: The guide compares proxy routes and backend plugins from an authentication perspective. Mentioning alerting here feels a little out-of-place? Maybe we could add a side note (using a blockquote) as a heads-up. 

@osg-grafana WDYT?

> - Proxy routes don't support custom token authentication

Reviewer 2: You're right

If any of these limitations apply to your plugin, you need to add a [backend plugin](TODO). Since backend plugins run on the server they can access decrypted secrets, which makes it easier to implement custom authentication methods.

The decrypted secrets are available from the `DecryptedSecureJSONData` field in the instance settings.

```go
func (ds *dataSource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {

>   instance, err := ds.im.Get(req.PluginContext)

Reviewer 2: This example uses the instance manager (im) but might be a bit unclear/confusing leaving that out in the example? 

This could have been nice to link to if it would have included usage of jsonData/secureJsonData

Author: I struggled to make an example for this that showcases the DecryptedSecureJsonData without having to include the bigger part of the plugin for context. What would be minimum amount of context we'd need to provide to demonstrate the specific use case of using the secure JSON data?

Reviewer 2: It's tricky. So perhaps just something like this could work?

```go
func (ds *dataSource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
  if mySecret, exists := req.PluginContext.DataSourceInstanceSettings.DecryptedSecureJSONData["mySecret"]; exists {
    // Call my custom authentication method using mySecret value...
  }
```

While the data source proxy supports the most common authentication methods for HTTP APIs, using proxy routes has a few limitations:

> - Proxy routes only support HTTP/HTTPS

Reviewer 3: ```suggestion
- Proxy routes only support HTTP or HTTPS
```
